### PR TITLE
[DeepSpeed] scale grad for zero-2

### LINKED ISSUE #3877: Gradient accumulation gives worse results when using DeepSpeed ZeRO 2
### System Info

```Shell
- `Accelerate` version: 1.11.0
- Platform: Linux-5.4.0-42-generic-x86_64-with-glibc2.35
- `accelerate` bash location: /raid/swyoon/miniconda3/envs/tbitrl/bin/accelerate
- Python version: 3.12.12
- Numpy version: 2.2.6
- PyTorch version: 2.7.1+cu126
- PyTorch accelerator: CUDA
- System RAM: 1006.58 GB
- GPU type: NVIDIA A100X
- `Accelerate` default config:
        Not found
```

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [x] My own task or dataset (give details below)

### Reproduction

In accelerate, the training with gradient accumulation gives higher training loss when DeepSpeed ZeRO stage 2, even though in theory the use of gradient accumulation should not affect the loss values.

For example, when experimented with an SFT script that is slightly modified from [TRL's script for testing gradient accumulation](https://github.com/huggingface/trl/issues/2635#issuecomment-2612759819), the loss curves are wildly different across gradient accumulation step values.
Image

Surprisingly, this issue does not happen when ZeRO stage 1 or 3 is used.

The issue can be reproduced by the following command.
```
CUDA_VISIBLE_DEVICES=0 accelerate launch --config_file ds_zero2.yaml --num_processes=1 sft_grad_accum_test.py
```
where `ds_zero2.yaml` and `sft_grad_accum_test.py` are given as follows.

sft_grad_accum_test.py:
```python
from datasets import load_dataset
from transformers import AutoTokenizer
from trl import SFTConfig, SFTTrainer
import os
os.environ["WANDB_PROJECT"] = "sft-grad-accum-test"

tokenizer = AutoTokenizer.from_pretrained("trl-internal-testing/tiny-Qwen2ForCausalLM-2.5")

batch_size = 2   # the result should be identical when batch_size=8 and gradient_accumulation_steps=1 are set.
gradient_accumulation_steps = 4
output_dir = f"SFT-bsz{batch_size}-grad_acc{gradient_accumulation_steps}-zero2"

training_args = SFTConfig(
    output_dir=output_dir,
    per_device_train_batch_size=batch_size,
    gradient_accumulation_steps=gradient_accumulation_steps,
    logging_steps=2,
    report_to="wandb",
    run_name=output_dir,
    completion_only_loss=True,
    learning_rate=1e-3,
)

dummy_dataset = load_dataset("trl-internal-testing/zen", "standard_prompt_completion")

trainer = SFTTrainer(
    model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
    args=training_args,
    processing_class=tokenizer,
    train_dataset=dummy_dataset["train"],
)

trainer.train()
```

ds_zero2.yaml
```
compute_environment: LOCAL_MACHINE
debug: false
deepspeed_config:
  deepspeed_multinode_launcher: standard
  offload_optimizer_device: none 
  offload_param_device: none
  zero3_init_flag: false
  zero_stage: 2
  gradient_clipping: auto
distributed_type: DEEPSPEED
downcast_bf16: 'no'
machine_rank: 0
main_training_function: main
mixed_precision: bf16
num_machines: 1
num_processes: 1
rdzv_backend: static
same_network: true
tpu_env: []
tpu_use_cluster: false
tpu_use_sudo: false
use_cpu: false
```

### Expected behavior

The training results should be identical as long as  `batch_size * gradient_accumulation_steps` remains constant, up to numerical precision.

For example, when `zero_stage` in `ds_zero2.yaml` is set to 1 or 3, the training curve looks as follows:

Image

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

This pull request updates the `backward` method in `src/accelerate/accelerator.py` to ensure consistent loss scaling across all distributed training backends.

Distributed training consistency:

* The loss is now always scaled by `gradient_accumulation_steps`, regardless of the backend, to prevent incorrect accumulation in DeepSpeed ZeRO-2 and similar scenarios. This change makes loss scaling explicit and consistent for all distributed types.

Fixes #3877

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/accelerate/blob/main/CONTRIBUTING.md#submitting-a-pull-request-pr),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/accelerate/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/accelerate/tree/main/docs#writing-documentation---specification).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -2840,9 +2840,12 @@ def backward(self, loss, **kwargs):
...
         learning_rate = kwargs.get("learning_rate")
 
-        if self.distributed_type != DistributedType.DEEPSPEED:
-            # deepspeed handles loss scaling by gradient_accumulation_steps in its `backward`
-            loss = loss / self.gradient_accumulation_steps
+        # Scale loss by gradient_accumulation_steps for all backends
+        # Note: DeepSpeed does NOT automatically scale loss during backward for all ZeRO stages,
+        # particularly ZeRO-2 where gradient partitioning can cause incorrect accumulation
+        # if the loss is not pre-scaled. This ensures consistent behavior across all ZeRO stages.
+        loss = loss / self.gradient_accumulation_steps
```
Comment (stas00) [inline src/accelerate/accelerator.py:2847]: Are you sure the issue isn't somewhere else?

You can see this is already done here:
https://github.com/deepspeedai/DeepSpeed/blob/b00b75f05852e0791f1e2b9c1cc894cd690e2da4/deepspeed/runtime/engine.py#L2482

and grads are scaled here:
https://github.com/deepspeedai/DeepSpeed/blob/b00b75f05852e0791f1e2b9c1cc894cd690e2da4/deepspeed/runtime/engine.py#L2360

so I suspect the above change is likely to break things, no? 

cc: @tjruwase to double check.

Comment (kashif) [reply src/accelerate/accelerator.py:2847]: yes agree! i was just wanted to test things... reverting

**Comment by kashif** [comment]
closing this
